File: sdks/elixir/lib/autonoma/create.ex
defmodule Autonoma.Create do
  @moduledoc "Create entities via raw SQL INSERT."

  @mysql_datetime_re ~r/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/

  def create_entities(executor, dialect, table_map, column_maps, spec, enum_type_maps \\ %{}, schema_models \\ []) do
    Enum.reduce(spec, %{}, fn {model, entity_spec}, acc ->
      db_table = Map.get(table_map, model) || raise "Unknown model \"#{model}\"."
      col_map = Map.get(column_maps, model, %{})
      enum_type_map = Map.get(enum_type_maps, model, %{})
      fields_list = Map.get(entity_spec, "fields", [])
      is_batch = Map.get(entity_spec, "batch", false)

      # Bug 4: find actual PK field name from schema
      model_info = Enum.find(schema_models, fn m -> m["name"] == model end)
      pk_field = if model_info, do: Enum.find(model_info["fields"] || [], fn f -> f["isId"] end)
      pk_field_name = if pk_field, do: pk_field["name"], else: "id"
      pk_field_type = if pk_field, do: pk_field["type"], else: "String"

      records =
        if is_batch && length(fields_list) > 0 do
          insert_batch(executor, dialect, db_table, col_map, enum_type_map, fields_list, pk_field_name, pk_field_type)
        else
          Enum.flat_map(fields_list, fn fields ->
            insert_one(executor, dialect, db_table, col_map, enum_type_map, fields, pk_field_name, pk_field_type)
          end)
        end

      Map.put(acc, model, records)
    end)
  end

  def update_entity(executor, dialect, table_map, column_maps, model, record_id, fields, enum_type_maps \\ %{}, pk_field_name \\ "id") do
    db_table = Map.get(table_map, model) || raise "Unknown model \"#{model}\" for update."
    col_map = Map.get(column_maps, model, %{})
    enum_type_map = Map.get(enum_type_maps, model, %{})

    {set_clauses, params, param_idx} =
      Enum.reduce(fields, {[], [], 1}, fn {field_name, value}, {clauses, ps, idx} ->
        db_col = Map.get(col_map, field_name, field_name)
        clause = "#{dialect.quote_id(db_col)} = #{cast_param(dialect, idx, enum_type_map, field_name)}"
        {clauses ++ [clause], ps ++ [serialize_value(value, dialect)], idx + 1}
      end)

    id_col = Map.get(col_map, pk_field_name, pk_field_name)
    sql = "UPDATE #{dialect.quote_id(db_table)} SET #{Enum.join(set_clauses, ", ")} WHERE #{dialect.quote_id(id_col)} = #{dialect.param(param_idx)}"
    executor.(:query, sql, params ++ [record_id])
  end

  # --- Internal ---

  defp insert_one(executor, dialect, db_table, col_map, enum_type_map, fields, pk_field_name \\ "id", pk_field_type \\ "String") do
    # Bug 1: Only generate UUID when PK type is String; Int/BigInt use DB auto-increment
    fields =
      if pk_field_name && !Map.has_key?(fields, pk_field_name) && pk_field_type == "String" do
        Map.put(fields, pk_field_name, UUID.uuid4())
      else
        fields
      end

    entries = Map.to_list(fields)

    if entries == [] do
      executor.(:query, "INSERT INTO #{dialect.quote_id(db_table)} DEFAULT VALUES RETURNING *", [])
      |> map_rows_back(col_map)
    else
      {db_cols, placeholders, params, _} =
        Enum.reduce(entries, {[], [], [], 1}, fn {field_name, value}, {cols, phs, ps, idx} ->
          db_col = Map.get(col_map, field_name, field_name)
          {
            cols ++ [dialect.quote_id(db_col)],
            phs ++ [cast_param(dialect, idx, enum_type_map, field_name)],
            ps ++ [serialize_value(value, dialect)],
            idx + 1
          }
        end)

      col_list = Enum.join(db_cols, ", ")
      val_list = Enum.join(placeholders, ", ")

      if dialect.supports_returning() do
        sql = "INSERT INTO #{dialect.quote_id(db_table)} (#{col_list}) VALUES (#{val_list}) RETURNING *"
        executor.(:query, sql, params) |> map_rows_back(col_map)
      else
        executor.(:query, "INSERT INTO #{dialect.quote_id(db_table)} (#{col_list}) VALUES (#{val_list})", params)
        id_col = Map.get(col_map, pk_field_name, pk_field_name)
        record_id = Map.get(fields, pk_field_name)
        executor.(:query, "SELECT * FROM #{dialect.quote_id(db_table)} WHERE #{dialect.quote_id(id_col)} = #{dialect.param(1)}", [record_id])
        |> map_rows_back(col_map)
      end
    end
  end

  defp insert_batch(_executor, _dialect, _db_table, _col_map, _enum_type_map, []), do: []

  defp insert_batch(executor, dialect, db_table, col_map, enum_type_map, fields_arr, pk_field_name \\ "id", pk_field_type \\ "String") do

    # Bug 1: Only generate client-side IDs when PK type is String; Int/BigInt use DB auto-increment
    fields_arr =
      if pk_field_name && pk_field_type == "String" do
        Enum.map(fields_arr, fn f ->
          if Map.has_key?(f, pk_field_name), do: f, else: Map.put(f, pk_field_name, UUID.uuid4())
        end)
      else
        fields_arr
      end

    field_names =
      fields_arr
      |> Enum.flat_map(&Map.keys/1)
      |> Enum.uniq()
      |> Enum.sort()

    if field_names == [] do
      Enum.flat_map(fields_arr, fn fields ->
        insert_one(executor, dialect, db_table, col_map, enum_type_map, fields)
      end)
    else
    db_cols = Enum.map(field_names, fn f -> dialect.quote_id(Map.get(col_map, f, f)) end)
    col_list = Enum.join(db_cols, ", ")

    # Chunk for bind var limits
    max_params = 32_000
    chunk_size = max(1, div(max_params, length(field_names)))

    fields_arr
    |> Enum.chunk_every(chunk_size)
    |> Enum.flat_map(fn chunk ->
      {tuples, params, _} =
        Enum.reduce(chunk, {[], [], 1}, fn fields, {ts, ps, idx} ->
          {phs, ps2, idx2} =
            Enum.reduce(field_names, {[], ps, idx}, fn fn_name, {ph_acc, p_acc, i} ->
              {
                ph_acc ++ [cast_param(dialect, i, enum_type_map, fn_name)],
                p_acc ++ [serialize_value(Map.get(fields, fn_name), dialect)],
                i + 1
              }
            end)
          {ts ++ ["(#{Enum.join(phs, ", ")})"], ps2, idx2}
        end)

      val_list = Enum.join(tuples, ", ")

      if dialect.supports_returning() do
        sql = "INSERT INTO #{dialect.quote_id(db_table)} (#{col_list}) VALUES #{val_list} RETURNING *"
        executor.(:query, sql, params) |> map_rows_back(col_map)
      else
        executor.(:query, "INSERT INTO #{dialect.quote_id(db_table)} (#{col_list}) VALUES #{val_list}", params)
        []
      end
    end)
    end
  end

  defp map_rows_back(rows, col_map) do
    reverse = Map.new(col_map, fn {field, db_col} -> {db_col, field} end)
    Enum.map(rows, fn row ->
      Map.new(row, fn {k, v} -> {Map.get(reverse, k, k), v} end)
    end)
  end

  defp cast_param(dialect, idx, enum_type_map, field_name) do
    placeholder = dialect.param(idx)
    if dialect.name() == "postgres" do
      case Map.get(enum_type_map, field_name) do
        nil -> placeholder
        enum_type -> "#{placeholder}::#{dialect.quote_id(enum_type)}"
      end
    else
      placeholder
    end
  end

  defp serialize_value(nil, _dialect), do: nil
  defp serialize_value(%DateTime{} = dt, dialect) do
    if dialect.name() == "mysql" do
      Calendar.strftime(dt, "%Y-%m-%d %H:%M:%S")
    else
      DateTime.to_iso8601(dt)
    end
  end
  # Bug 2: Only JSON-encode maps/dicts, NOT lists. Return lists as native values for Postgres ARRAY columns.
  defp serialize_value(value, _dialect) when is_map(value), do: Jason.encode!(value)
  defp serialize_value(value, _dialect) when is_list(value), do: value
  defp serialize_value(value, dialect) when is_binary(value) do
    if dialect.name() == "mysql" && Regex.match?(@mysql_datetime_re, value) do
      value |> String.replace("T", " ") |> String.replace("Z", "") |> String.replace(~r/\.\d+$/, "")
    else
      value
    end
  end
  defp serialize_value(value, _dialect), do: value
end

# Minimal UUID generation (no external dep needed)
defmodule UUID do
  @moduledoc false
  def uuid4 do
    <<a::32, b::16, c::16, d::16, e::48>> = :crypto.strong_rand_bytes(16)
    :io_lib.format("~8.16.0b-~4.16.0b-~4.16.0b-~4.16.0b-~12.16.0b", [a, b, c, d, e])
    |> to_string()
  end
end

File: sdks/go/autonoma/create.go
package autonoma

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/google/uuid"
)

// CreateEntities inserts entities via raw SQL. Entities arrive pre-sorted by FK order.
func CreateEntities(
	ctx context.Context,
	executor SQLExecutor,
	dialect *Dialect,
	tableMap map[string]string,
	columnMaps map[string]map[string]string,
	spec map[string]ResolvedEntitySpec,
	enumTypeMaps map[string]map[string]string,
) (map[string][]map[string]any, error) {
	results := make(map[string][]map[string]any)

	for model, entitySpec := range spec {
		dbTable, ok := tableMap[model]
		if !ok {
			return nil, fmt.Errorf("unknown model %q. Not found in database tables", model)
		}
		colMap := columnMaps[model]
		if colMap == nil {
			colMap = make(map[string]string)
		}
		enumTypeMap := enumTypeMaps[model]
		if enumTypeMap == nil {
			enumTypeMap = make(map[string]string)
		}

		var created []map[string]any
		for _, fields := range entitySpec.Fields {
			records, err := insertOne(ctx, executor, dialect, dbTable, colMap, enumTypeMap, fields)
			if err != nil {
				return nil, err
			}
			if len(records) > 0 {
				created = append(created, records[0])
			}
		}
		results[model] = created
	}

	return results, nil
}

// UpdateEntity updates a single record by primary key. Used for circular FK backfill.
func UpdateEntity(
	ctx context.Context,
	executor SQLExecutor,
	dialect *Dialect,
	tableMap map[string]string,
	columnMaps map[string]map[string]string,
	model string,
	id string,
	fields map[string]any,
	enumTypeMaps map[string]map[string]string,
) error {
	dbTable, ok := tableMap[model]
	if !ok {
		return fmt.Errorf("unknown model %q for update", model)
	}
	colMap := columnMaps[model]
	if colMap == nil {
		colMap = make(map[string]string)
	}
	enumTypeMap := enumTypeMaps[model]
	if enumTypeMap == nil {
		enumTypeMap = make(map[string]string)
	}

	var setClauses []string
	var params []any
	paramIdx := 1

	for fieldName, value := range fields {
		dbCol := colMap[fieldName]
		if dbCol == "" {
			dbCol = fieldName
		}
		setClauses = append(setClauses, fmt.Sprintf("%s = %s", dialect.QuoteID(dbCol), castParam(dialect, paramIdx, enumTypeMap, fieldName)))
		params = append(params, serializeValue(value, dialect))
		paramIdx++
	}

	idCol := colMap["id"]
	if idCol == "" {
		idCol = "id"
	}
	params = append(params, id)

	sql := fmt.Sprintf("UPDATE %s SET %s WHERE %s = %s",
		dialect.QuoteID(dbTable),
		strings.Join(setClauses, ", "),
		dialect.QuoteID(idCol),
		dialect.Param(paramIdx))

	_, err := executor.Query(ctx, sql, params...)
	return err
}

func insertOne(
	ctx context.Context,
	executor SQLExecutor,
	dialect *Dialect,
	dbTable string,
	colMap map[string]string,
	enumTypeMap map[string]string,
	fields map[string]any,
) ([]map[string]any, error) {
	// Generate client-side ID if needed
	idFieldName := reverseGetMap(colMap, findIDCol(colMap))
	if idFieldName != "" {
		if _, exists := fields[idFieldName]; !exists {
			fieldsCopy := make(map[string]any, len(fields)+1)
			for k, v := range fields {
				fieldsCopy[k] = v
			}
			fieldsCopy[idFieldName] = uuid.New().String()
			fields = fieldsCopy
		}
	}

	// Remove "id" key if it exists as nil
	cleanFields := make(map[string]any)
	for k, v := range fields {
		if k == "id" && v == nil {
			continue
		}
		cleanFields[k] = v
	}
	fields = cleanFields

	if len(fields) == 0 {
		rows, err := executor.Query(ctx, fmt.Sprintf("INSERT INTO %s DEFAULT VALUES RETURNING *", dialect.QuoteID(dbTable)))
		if err != nil {
			return nil, err
		}
		return mapRowsBack(rows, colMap), nil
	}

	var dbCols []string
	var params []any
	var placeholders []string
	paramIdx := 1

	for fieldName, value := range fields {
		dbCol := colMap[fieldName]
		if dbCol == "" {
			dbCol = fieldName
		}
		dbCols = append(dbCols, dialect.QuoteID(dbCol))
		placeholders = append(placeholders, castParam(dialect, paramIdx, enumTypeMap, fieldName))
		params = append(params, serializeValue(value, dialect))
		paramIdx++
	}

	colList := strings.Join(dbCols, ", ")
	valList := strings.Join(placeholders, ", ")

	if dialect.SupportsReturning {
		sql := fmt.Sprintf("INSERT INTO %s (%s) VALUES (%s) RETURNING *",
			dialect.QuoteID(dbTable), colList, valList)
		rows, err := executor.Query(ctx, sql, params...)
		if err != nil {
			return nil, err
		}
		return mapRowsBack(rows, colMap), nil
	}

	// MySQL: INSERT then SELECT
	sql := fmt.Sprintf("INSERT INTO %s (%s) VALUES (%s)",
		dialect.QuoteID(dbTable), colList, valList)
	_, err := executor.Query(ctx, sql, params...)
	if err != nil {
		return nil, err
	}

	idCol := findIDCol(colMap)
	idValue := fields[idFieldName]
	if idFieldName == "" {
		idValue = fields["id"]
	}

	selectSQL := fmt.Sprintf("SELECT * FROM %s WHERE %s = %s",
		dialect.QuoteID(dbTable), dialect.QuoteID(idCol), dialect.Param(1))
	rows, err := executor.Query(ctx, selectSQL, idValue)
	if err != nil {
		return nil, err
	}
	return mapRowsBack(rows, colMap), nil
}

func mapRowsBack(rows []map[string]any, colMap map[string]string) []map[string]any {
	if len(colMap) == 0 {
		return rows
	}

	reverse := make(map[string]string, len(colMap))
	for fieldName, dbCol := range colMap {
		reverse[dbCol] = fieldName
	}

	result := make([]map[string]any, len(rows))
	for i, row := range rows {
		mapped := make(map[string]any, len(row))
		for key, value := range row {
			fieldName := reverse[key]
			if fieldName == "" {
				fieldName = key
			}
			mapped[fieldName] = value
		}
		result[i] = mapped
	}
	return result
}

func findIDCol(colMap map[string]string) string {
	if v, ok := colMap["id"]; ok {
		return v
	}
	return "id"
}

func castParam(dialect *Dialect, paramIdx int, enumTypeMap map[string]string, fieldName string) string {
	placeholder := dialect.Param(paramIdx)
	if dialect.Name == "postgres" {
		if enumType, ok := enumTypeMap[fieldName]; ok {
			return placeholder + "::" + dialect.QuoteID(enumType)
		}
	}
	return placeholder
}

var isoDateTimeRE = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}`)

func serializeValue(value any, dialect *Dialect) any {
	if value == nil {
		return value
	}

	// Handle time.Time
	if t, ok := value.(time.Time); ok {
		if dialect.Name == "mysql" {
			return t.Format("2006-01-02 15:04:05")
		}
		return t.Format(time.RFC3339Nano)
	}

	// JSON: maps and slices need to be stringified
	switch v := value.(type) {
	case map[string]any:
		b, _ := json.Marshal(v)
		return string(b)
	case []any:
		b, _ := json.Marshal(v)
		return string(b)
	}

	// MySQL datetime string conversion
	if s, ok := value.(string); ok && dialect.Name == "mysql" {
		if isoDateTimeRE.MatchString(s) {
			s = strings.Replace(s, "T", " ", 1)
			s = strings.Replace(s, "Z", "", 1)
			// Remove fractional seconds
			if idx := strings.LastIndex(s, "."); idx > 0 {
				s = s[:idx]
			}
			return s
		}
	}

	return value
}
